upgrade/scripts/upgrade/util/fix-postgres.sh

```diff
@@ -54,12 +54,26 @@ function postgres_crd_updated() {
   fi
 }
 
-function postgres_pods_running() {
-  if [ "$(kubectl get pod -l application=spilo -A --no-headers | awk '{ print $3 ":" $4 }' | sort -u)" == "3/3:Running" ]; then
-    return 0
-  else
-    return 1
-  fi
+# Query the statefulsets status, replicas = goal, readyReplicas = how many pods
+# are receiving traffic.
+#
+# Iff replicas != readyReplicas the cluster isn't ready.
+#
+# We do this so as not to hard code expectations on how many replicas might be
+# configured for the pg cluster.
+function postgres_statefulsets_ready() {
+  for ns in $(kubectl get ns --no-headers -o custom-columns=":metadata.name"); do
+    for ss in $(kubectl get statefulset -l application=spilo --namespace "${ns}" --no-headers -o custom-columns=":metadata.name"); do
+      # Early break on first statefulset not fully ready.
+      # Note ${ns} and ${ss} are already in a "" block, this shellcheck warning is pointless here.
+      #shellcheck disable=SC2086
+      if [ "$(kubectl get statefulset --namespace ${ns} ${ss} -o jsonpath='{.status.replicas}{"\n"}{.status.readyReplicas}{"\n"}' | sort -u | wc -l)" -ne 1 ]; then
+        return 1
+      fi
+    done
+  done
+
+  return 0
 }
```

Comment:
That presumes all the ready and replicas are equal for all statefulsets which isn't true as we have 3 and 2 respectively. That is the underlying reason for individual calls.

Reply:
No, it doesn't presume anything.  It gets a list of all the ready/replicas values in one call instead of doing a `kubectl` call for each pair.  It then checks each of them looking to see if any of them don't match (the `| sort -u` is a minor optimization that isn't really needed).  It's doing the exact same thing as your code.

Reply:
For example, on mug right now:

Reply:
My apologies, I didn't fully check that the return value was correct.  With the `|` it was running in a subshell, so the failure wasn't getting returned properly.  Here's the version that doesn't use a subshell:
```
function postgres_statefulsets_ready() {
    while read rep ready; do
      if [ "$rep" != "$ready" ]; then
        return 1
      fi
    done < <(kubectl get statefulset -A -l application=spilo  -o jsonpath='{range .items[*]} {.status.replicas} {.status.readyReplicas}{"\n"}{end}')
  return 0
}
```

Reply:
Please resolve the conversation/thread if you're done.

Reply:
I would like to see the code changed to a more efficient version, here's a timing comparison:

Reply:
@mitchty, though I would prefer that this PR be changed to the more efficient code that only makes one `kubectl` call, this is not code that is run on a regular basis so I won't insist on this change.  You can choose to either make the change or add a comment that you aren't going to change it (since this is an infrequently used code path).  In either case, then mark this conversation as `resolved` and the merge will be able to proceed.

Reply:
Fair points, though this only runs once on an upgrade so I wasn't too concerned with runtime. My only apprehension is if something is broken or not in Ready we'd need to do the same to track down which statefulset is at issue.

Let me test this out after breaking a vshasta setup quick.

Reply:
Yes, if you want to print out which ones aren't ready, this will do that (and go through the whole list, not just stop at the first that isn't ready).
```
function postgres_statefulsets_ready() {
  ret=0
  while read -r ns ss rep ready; do
    if [ "$rep" != "$ready" ]; then
      ret=1
      echo "Stateful set $ns/$ss not ready: replicas: $rep, ready: $ready"
    fi
  done < <(kubectl get statefulset -A -l application=spilo  -o jsonpath='{range .items[*]}{.metadata.namespace} {.metadata.name} {.status.replicas} {.status.readyReplicas}{"\n"}{end}')
  return $ret
}
```

Reply:
Should be updated with an autosquash! commit to basically do the suggestion. Tested it out with another broken (by me) virtual cluster.

Reply:
Thanks!  That looks good to me.